src/wp-includes/blocks.php

```diff
@@ -1125,12 +1125,34 @@ function apply_block_hooks_to_content( $content, $context = null, $callback = 'i
 		}
 	}
 
-	/*
-	 * We also need to cover the case where the hooked block is not present in
-	 * `$content` at first and we're allowed to insert it once -- but not again.
-	 */
-	$suppress_single_instance_blocks = static function ( $hooked_block_types ) use ( &$block_allows_multiple_instances, $content ) {
+	$suppress_blocks_from_insertion = static function ( $hooked_block_types, $relative_position, $anchor_block_type ) use ( &$block_allows_multiple_instances, $content, $context ) {
 		static $single_instance_blocks_present_in_content = array();
+
+		/*
+		 * If the context is a post object, we need to avoid inserting any blocks hooked into the
+		 * `before` and `after` positions of the temporary wrapper block that we create to wrap the content.
+		 * See https://core.trac.wordpress.org/ticket/63287 for more details.
+		 */
+		if ( $context instanceof WP_Post ) {
+			$wrapper_block_type = 'core/post-content';
+			if ( 'wp_navigation' === $context->post_type ) {
+				$wrapper_block_type = 'core/navigation';
+			} elseif ( 'wp_block' === $context->post_type ) {
+				$wrapper_block_type = 'core/block';
+			}
+
+			if (
+				$wrapper_block_type === $anchor_block_type &&
+				in_array( $relative_position, array( 'before', 'after' ), true )
+			) {
+				return array();
+			}
+		}
```

Comment:
I'm wondering if we should move this logic over to `apply_block_hooks_to_content_from_post_object`, as that is the function that includes all the wrapper block logic, whereas `apply_block_hooks_to_content` is otherwise agnostic of that. So from a separation-of-concerns POV, it might make sense.

The downside is that that would require having a separate (anonymous) filter function declared inside of `apply_block_hooks_to_content_from_post_object` that is then also added to the `hooked_block_types`, much like we're doing here already.

Reply:
I'm going to give this idea a quick try...

Reply:
Done in https://github.com/WordPress/wordpress-develop/pull/8700/commits/987436a83a2020433b7c3fb382a4534f367f4428 and https://github.com/WordPress/wordpress-develop/pull/8700/commits/7aa3df744a4afc18f01267fc7267c05ecdbeeef6. Looks better this way IMO.

Reply:
That's a nice refactoring to move it to the logic in a different place where awareness for the block wrapper is already in place 👍🏻